### DISCUSSION TITLE
Add compiler flag to prevent lambda parameters from hiding names from the enclosing scope

### DISCUSSION THREAD

**Discussion by Timwi** [dotnet/csharplang#5351] [Language Ideas]
I personally feel that allowing lambda expressions to hide names from enclosing scopes (added in C# 8) was a mistake.

I’ve only started using the new C# language features for a few months and I’ve already created two bugs this way that were not caught by the compiler, while they previously would have been. In one case, I accidentally hid a for-loop variable, in the other it was another parameter from an outer nested lambda.

I’ve noticed [this discussion](https://github.com/dotnet/csharplang/issues/2777) in which it is believed that the ability to hide names “will decrease developer frustration, with only limited chance of causing bugs”. My experience suggests this to be quite misguided. I’ve experienced _more_ bugs and, as a result, more frustration trying to track them down.

Since this change was made in C# 8.0 and C# 9.0 has already shipped, it’s obviously no longer possible to reverse this without breaking existing code. For this reason, I would like to propose to add a compiler switch/build option, which would allow project maintainers to use the previous behavior, which catches these types of bugs.

Thank you for listening.

**Comment by 333fred** [comment]
An analyzer would be able to warn or error on this case, without having to change the language. It's a prime use case for one.
